blob/src/main/java/io/crate/blob/v2/BlobIndex.java

```diff
@@ -65,24 +71,71 @@ public static String stripPrefix(String indexName) {
 
     private final Map<Integer, BlobShard> shards = new ConcurrentHashMap<>();
     private final Path globalBlobPath;
+    private final ESLogger logger;
 
-
-    BlobIndex(@Nullable Path globalBlobPath) {
+    BlobIndex(ESLogger logger, @Nullable Path globalBlobPath) {
         this.globalBlobPath = globalBlobPath;
+        this.logger = logger;
     }
 
     void createShard(IndexShard indexShard) {
         shards.put(indexShard.shardId().id(), new BlobShard(indexShard, globalBlobPath));
     }
 
     BlobShard removeShard(ShardId shardId) {
+        Path blobRoot = null;
         BlobShard shard = shards.remove(shardId.id());
         if (shard != null) {
+            if (shards.size() == 0) {
+                blobRoot = this.retrieveBlobRootDir(shard);
+            }
             shard.deleteShard();
+            if (blobRoot != null) {
+                this.deleteIndex(blobRoot, shardId.getIndex());
+            }
         }
+
         return shard;
     }
 
+    /**
+     * Traverse the path down until the shard index was found
+     * and return the root path of the blob directory.
+     */
+    private Path retrieveBlobRootDir(BlobShard blobShard) {
+        Path blobDir = blobShard.getBlobDir();
+        String indexName = blobShard.indexShard().shardId().getIndex();
+        Path resultDir = Paths.get(blobDir.toString());
+        for (Iterator i = blobDir.iterator(); i.hasNext();) {
```

Comment:
I had something like ` while ((resultDir = resultDir.getParent()) != null) { indexNameCheck }` in mind - the iterator now seems unused but it still requires some form of loop?